test/DAICO.test.js

```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
+      assert.ok(new_tap.c[0]==100);
+      await this.daico.newRaiseTapProposal(99, 3600, {from: purchaser}).should.be.rejectedWith(EVMRevert);
+    });
+
+    it('old propose can not be executed if it lowers the tap', async function() {
+      let purchaser = accounts[4];
+      await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value});
+      await increaseTimeTo(this.afterlastWithdrawn);
+      await this.daico.newRaiseTapProposal(100, 3600, {from: purchaser});
+      await this.daico.newRaiseTapProposal(90, 7200, {from: purchaser});
+      await this.daico.vote(0, true, {from: purchaser}).should.be.fulfilled;;
+      await this.daico.vote(1, true, {from: purchaser}).should.be.fulfilled;;
+      await increaseTimeTo(latestTime() + duration.hours(2) + duration.seconds(1));
+      await this.daico.executeRaiseTapProposal(0).should.be.fulfilled;
+      await this.daico.executeRaiseTapProposal(1).should.be.rejectedWith(EVMRevert);
+      let new_tap = await this.daico.tap();
+      assert.ok(new_tap.c[0]==100);
+    });
   });
 
+  describe('tap', function () {
+
+    it('owner cant withdraw before last withdrawal', async function () {
+      let owner = accounts[0];
+      let no_owner = accounts[3];
+      await this.daico.buyTokens(no_owner, {from: no_owner, value: higher_value}).should.be.fulfilled;
+      await this.daico.withdraw({from: owner}).should.be.rejectedWith(EVMRevert);
+    });
+
+    it('initial tap is zero and owner can withdraw zero', async function () {
+      let owner = accounts[0];
+      let no_owner = accounts[3];
+      await this.daico.buyTokens(no_owner, {from: no_owner, value: higher_value}).should.be.fulfilled;
+      await increaseTimeTo(this.afterlastWithdrawn);
+      const initial_balance = web3.eth.getBalance(owner).c[0];
+      await this.daico.withdraw({from: owner}).should.be.fulfilled;
+      const final_balance = web3.eth.getBalance(owner).c[0];
+      assert.ok(final_balance < initial_balance);
+    });
+
+    it('owner can withdraw as much as tap times time', async function () {
+      let owner = accounts[0];
+      let no_owner = accounts[3];
+      await this.daico.buyTokens(no_owner, {from: no_owner, value: higher_value}).should.be.fulfilled;
+      await increaseTimeTo(this.afterlastWithdrawn);
+      await this.daico.newRaiseTapProposal(tap, 3600, {from: no_owner});
+      await this.daico.vote(0, true, {from: no_owner}).should.be.fulfilled;;
+      await increaseTimeTo(latestTime() + duration.hours(1) + duration.seconds(1));
+      await this.daico.executeRaiseTapProposal(0).should.be.fulfilled;
+      await increaseTimeTo(latestTime() + duration.seconds(1000000));
+      const initial_balance = web3.eth.getBalance(owner).c[0];
+      const out = await this.daico.withdraw({from: owner});
+      const fee = out.receipt.gasUsed / 1000;
+      const final_balance = web3.eth.getBalance(owner).c[0];
+      assert.ok(final_balance - initial_balance - 100 + fee < 1);
+    });
+
+  });
```

Comment:
There is a lot repeated code to setup some of these tests, feel free to use a `beforeEach` statement to reduce it

Reply:
This is the first thing that I did. That is why commit `c185a55` is big, sorry for that.